README.md
msaccess-vcs-integration
========================

About
-----

Easily export your Microsoft Access Database objects for use with a version control system like **GitHub** or **GitLab**. (Allows you to easily track code changes over time.)

Note that this branch has taken **a significantly different direction** from the original (and primary) project maintained at <https://github.com/timabell/msaccess-vcs-integration>.

This project works as a **Microsoft Access add-in**, loaded on demand from the VBE IDE, and is designed to work with ADP projects as well as the main accdb format. This project is under development for internal use, but I will try to push updates to GitHub as I am able.

Another key difference with this project is that my focus has been strictly on the *export* of database objects into a version control system like GitHub or GitLab for the purpose of tracking changes to a project over time.

This project also supports **Microsoft Access ADP projects**, a now-deprecated flavor of an Access database that integrates more tightly with SQL Server. A few of us still support such projects, and this tool makes that job easier.  :-) 

Installation
------------

To use this project, you will need the following two things:
 1. Copy this add-in file (**Version Control.accda**) into the same folder as your Microsoft Access database.
 2. Then, copy or import **`modVersionControl`** module from `Version Control.accda` into the database you want to track. The defaults should work right out of the box, but you can see below for more advanced usage.
 
 ![version-control-module](img/version-control-module.jpg)

 Basic Usage
 -----------
 To use this add-in, simply type `InitializeVersionControl` from the VBA immediate window. (Ctl+space is a great auto-completion shortcut to use with this.)
 
 ![initialize](img/initialize.jpg)

 
 This will load a toolbar with a button to **Export All**. Click this button to export the source code and objects from the current database. 
 
![export-all](img/export-all.jpg)

You will then see some text output in your immediate window as the tool exports the database objects. (See advanced options if you want to see each object listed as it is exported.) You will see the number of each type of object, as well as the total time it took to run.

![sample-output](img/sample-output.jpg)


Advanced Options
----------------

The basic concept here is that the `modVersionControl` module is a loader that loads the MSAccessVCS library into the current project as a referenced library add-in. If you don't want it loaded all the time, you can run `InitializeVersionControl False` to unload the add-in module after you are finished.

### VCS Settings

Inside `modVersionControl`, you will see a number of options that you can customize for your needs. The names are pretty self-explanatory, but here is a little more description on what they do:

 * **System** - I primarily use GitHub, (which also works for GitLab) but there is some support for TortoiseSVN, which I have used in the past.
 * **Export Folder** - I like to keep this relative to the project, but it can be customized here.
 * **Show Debug** - Show more verbose messages while running.
 * **Include VBE** - Access objects are always exported, and include the VBA code. This option gives you the ability to also export the VBE objects, which can be helpful if you are dragging and dropping these files to other projects, or importing as VBE objects.
 * **Fast Save** - Major performance gain with small changes to large projects. This attempts to only export the objects that have changed since the last export. This especially helps to not have to export forms and reports if they have not changed.
 * **Save Print Vars** - Saves a copy of the print variables, such as custom printer settings used in reports.
 * **Save Query SQL** - In addition to the Access object, this option exports a copy of just the SQL code from queries. I find this much more readable than the source of the Access Object when reviewing what I actually changed on the SQL side. (The Access object includes other information relating to the layout of the query designer.)
 * **Save Table SQL** - In addition to the Access object, this creates a SQL statement like what you would use to create the table. Here again I find this easier to see at a glance what changed in the actual structure of the table between versions.
 * **Save Table** - Includes a copy of the table data in the export. Normally you are just exporting the object definitions, but in some cases you may want data from some of the tables as well. Repeat this option for each table for which you want to save data.
    
```vb    '---------------------------------------------------------------------------------------
' Procedure : InitializeVersionControlSystem
' Author    : Adam Waller
' Date      : 1/29/2019
' Purpose   : Call this function to activate or deactivate version control.
'           : Add a copy of this module to the local project to serve as the
'           : loader/unloader of the referenced VCS library.
'---------------------------------------------------------------------------------------
'
Public Sub InitializeVersionControlSystem(Optional blnUseVersionControl As Boolean = True)

    '//////////////////////////////////////////////////////////
    ' SET THESE VALUES AS NEEDED FOR YOUR ENVIRONMENT
    '//////////////////////////////////////////////////////////
    
    ' Version Control Library
    Const cstrLibraryPath As String = "\"
    Const cstrLibraryFile As String = "Version Control.accda"
    Const cstrLibraryName As String = "MSAccessVCS"
    
    ' VCS Settings for this database (Additional parameters may be added as needed)
    Dim colParams As New Collection
    With colParams
        .Add Array("System", "GitHub")  ' IMPORTANT: Set this first, before other settings.
        .Add Array("Export Folder", CurrentProject.Path & "\" & CurrentProject.Name & ".src\")
        ' Optional parameters
        .Add Array("Show Debug", False)
        .Add Array("Include VBE", False)
        .Add Array("Fast Save", True)
        .Add Array("Save Print Vars", False)
        .Add Array("Save Query SQL", True)
        .Add Array("Save Table SQL", True)
        '.Add Array("Save Table", "vObjectPermissions")
    End With

    ' Make sure we are in the current project, not a library database.
    If CurrentProject.FullName <> CodeProject.FullName Then
        ' Call the function in the current project instead.
        Debug.Print "Please run this command from " & CurrentProject.Name
    Else
        ' Pass the parameters to the wrapper function
        LoadVersionControl blnUseVersionControl, cstrLibraryPath, cstrLibraryFile, cstrLibraryName, colParams
    End If

End Sub
```

### Automatic Loading

To load this library automatically when your database opens, include the following line in your startup code:

`InitializeVersionControlSystem`

### Renaming Loader Module

You are free to rename the loader module to whatever you like to conform to your project's naming convention. (I.e. you might prefer `basVersionControl` instead of `modVersionControl`)

CHANGELOG.md
Change log
==========
Ongoing Development
-------------------
joyfullservice:

As this project continues to move forward, many changes and updates are being made but not specifically documented in this change log. For details on updates since 2015, please review the commit messages in this repository.

(If you use this project and would find more detailed change logs to be helpful, please let me know, and I will take that into consideration.)

Version 1.1.0 - 15 May 2015
----------------------------
joyfullservice:
* Performed some significant refactoring to use the system as a reference library rather than a collection of modules that need to be added to each project.
* Added support for exporting VBE objects directly, allowing drag-and-drop within projects in the IDE.
* Standardized module names (special prefix no longer needed)
* Source folder name changed to be specific to the current database (avoids unexpected results when working with two databases in the same folder.)
* Source folders only created when they will contain content
* General code cleanup and organization
* Improved debug output in debug mode
* Converted module constants to input parameters for debug mode and tables to save data
* Updated sample database

Version 1.0.0 - 11 Mar 2015
----------------------------
jwbrookes:
* Added support for Table Data Macros
* Added support for Linked Tables (supports relative paths for linked files)
* Added support for Print Variables in Reports (Page size and orientation) 
* Added support for Relation for all types of table
* LoadVCS warnings removed when no object delete is required
* Removed elements from Report export that change constantly but don't affect import
* Fixed query import bug (complex queries being rearranged on import)
* Fixed missing constraints in table export

prmills:
* Added support for bit fields in table Import/Export
* Added support for References without GUIDs
* Export all table data with `INCLUDE_TABLES = "*"`

Version 0.12.0 - 28 Jan 2015
----------------------------
jwbrookes:
* Refactored AppCodeImportExport into several modules
* Added VCS_Loader, a module to import multiple vba modules into a database 


Version 0.11.1 - 14 Jan 2015
---------------------------
jwbrookes:
* Fixed bug in ExportTableDef function
* Removed redundant DeleteFile function (had been left commented out in the module)


Version 0.11 - 01 May 2014
-------------------------
matonb:
* Removed DeleteFile function and replaced calls to Kill with FileSystemObject.DeleteFile


Version 0.10 - 09 Mar 2014
-------------------------
matonb:
* Added DoEvents in loops to avoid "Unresponsive" state.


Version 0.9 - 15 Feb 2014
-------------------------
matonb:
*  Aggressive Sanitise, moved BaseInfo from "Block" regex to Line level.
*  Changed line level skipping to include lines with deeper indendation the follow.
  This catches split lines mostly found in BaseInfo exports.


Version 0.8 - 14 Feb 2014
-------------------------
matonb:
*  Aggressive Sanitise now excludes "BaseInfo" lines.
  These lines were seen to be randomly switching between being empty,
  not present or containing SQL on an arbitary basis.


Version 0.7 - 06 Jul 2013
-------------------------
matonb:
*  Replaced TempFile function.
*  Temporary file names now generated via external MS libraries.
*  Functions using TempFile updated to only call TempFile function once.
   *  Temporary file path and name stored in tmepFileName variable.
   *  Temporary files deleted when done.
*  Changed db declaration in ImportProject to DAO.database.


Version 0.6 - 06 Jul 2013
-------------------------

matonb:

*  AppcodeImportExport excluded from ExportAllSource
*  Added ImportProject sub-routine,  
   Deletes all forms, macros, modules and queries before calling ImportAllSource.  
   By clearing out the existing objects, you know that your database only contains  
   code from your version control database.  
   Excludes *AppCodeImportExport*

Version 0.5 - 29 May 2013
--------------------------

matonb:

*  All "exclusion" patterns are now matched by regex.
*  Added StripPublishOption constant.  
   If set to _True_ the following lines are also excluded from the export files
  * dbByte "PublishToWeb" ="1"
  * PublishOption =1
*  Added DeleteFile(FileName) function  
   The function tries to delete _FileName_ three (3) times before giving up.  
   A delay of 100ms is introduced between delete attempts should the first fail.

Version 0.4 - 19 Apr 2013
--------------------------

matonb:

*  Added dbLongBinary "DOL" to aggressive sanitize, these statements were
   appearing in queries and being flagged by git as modified in files that
   hadn't been touched by developers.

Version 0.3.2 - 8 Apr 2013
--------------------------

matonb:
* 0.3.1 Patched - Serious Problem:  SanitizeTextFiles If logic removed all
        lines containing "Begin".
* 0.3.2 Replaced if block for skipping code sections in SanitizeTextFiles with
        regular expression.

Version 0.3 - 6 Apr 2013
------------------------

bkidwell:
* Sanitize query exports.
* Fixed SERIOUS TYPO in UCS2-to-UTF-8 conversion (wrong threshold for 2 byte versus 3 byte symbol in output stream).
* AggressiveSanitize default True.

matonb:
* Added AggressiveSanitize constant, it's a number to allow for different levels in the future. ~~Default False.~~
* Added Skipping for GUID & Namemap in aggressive sanitize mode.
* ~~If AggressiveSanitize is on, also sanitize query exports.~~
* Append Number of objects imported/exported to information lines in immediate window.
* Updated readme (removed references to terminal window).
* Close all open forms and reports when importing and exporting because you can't import an open form or report.

Version 0.2 - 4 Apr 2013
------------------------

matonb:
* Added dbLongBinary "DOL" to SkipList in SanitizeTextFiles.
* Added Source directory check to ImportAllSource, pops up a message box if missing.

LoadVersionControl.bas
Option Explicit
Option Private Module
Option Compare Database


'---------------------------------------------------------------------------------------
' Procedure : InitializeVersionControlSystem
' Author    : Adam Waller
' Date      : 1/29/2019
' Purpose   : Call this function to activate or deactivate version control.
'           : Add a copy of this module to the local project to serve as the
'           : loader/unloader of the referenced VCS library.
'---------------------------------------------------------------------------------------
'
Public Sub InitializeVersionControlSystem(Optional blnUseVersionControl As Boolean = True)

    '//////////////////////////////////////////////////////////
    ' SET THESE VALUES AS NEEDED FOR YOUR ENVIRONMENT
    '//////////////////////////////////////////////////////////
    
    ' Version Control Library
    Const cstrLibraryPath As String = "\"
    Const cstrLibraryFile As String = "Version Control.accda"
    Const cstrLibraryName As String = "MSAccessVCS"
    
    ' VCS Settings for this database (Additional parameters may be added as needed)
    Dim colParams As New Collection
    With colParams
        .Add Array("System", "GitHub")  ' IMPORTANT: Set this first, before other settings.
        .Add Array("Export Folder", CurrentProject.Path & "\" & CurrentProject.Name & ".src\")
        ' Optional parameters
        .Add Array("Show Debug", False)
        .Add Array("Include VBE", False)
        .Add Array("Fast Save", True)
        .Add Array("Save Print Vars", False)
        .Add Array("Save Query SQL", True)
        .Add Array("Save Table SQL", True)
        '.Add Array("Save Table", "vObjectPermissions")
    End With

    ' Make sure we are in the current project, not a library database.
    If CurrentProject.FullName <> CodeProject.FullName Then
        ' Call the function in the current project instead.
        Debug.Print "Please run this command from " & CurrentProject.Name
    Else
        ' Pass the parameters to the wrapper function
        LoadVersionControl blnUseVersionControl, cstrLibraryPath, cstrLibraryFile, cstrLibraryName, colParams
    End If

End Sub


'---------------------------------------------------------------------------------------
' Procedure : LoadVersionControl
' Author    : Adam Waller
' Date      : 5/15/2015
' Purpose   : Attempt to relink the MSAccess-VCS library from the current folder
'           : if the reference is reported as broken. This allows you to either specify
'           : a location that stays the same across projects, or to use the relative
'           : path of the current folder.
'---------------------------------------------------------------------------------------
'
Private Sub LoadVersionControl(blnUseVersionControl As Boolean, strLibraryPath As String, strLibraryFile As String, strLibraryName As String, colParams As Collection)
    
    Const VB_PROJECT As Integer = 1

    Dim ref As Reference
    Dim intCnt As Integer
    Dim strPath As String
    Dim strFile As String
    Dim varParts As Variant
    Dim blnInitialize As Boolean
    Dim blnLoaded As Boolean
    Dim strLibPrefix As String
    Dim strError As String
    
    ' Unable to use library name when loading from an ADP project.
    If CurrentProject.ProjectType <> acADP Then strLibPrefix = "[" & strLibraryName & "]."
        
    ' Loop backwards through references, since libraries will be near the end.
    For intCnt = Application.References.Count To 1 Step -1
        Set ref = Application.References(intCnt)
        If ref.Kind = VB_PROJECT Then
            If ref.Name = strLibraryName Then
                If blnUseVersionControl Then
                    ' Make sure the path is valid.
                    If Dir(ref.FullPath) = "" Then
                        ' Could not find at current path. Get file name
                        varParts = Split(ref.FullPath, "\")
                        strFile = varParts(UBound(varParts))
                        ' Check in current folder
                        strPath = CodeProject.Path & "\" & strFile
                        If Dir(strPath) <> "" Then
                            ' Found in current folder. Relink at this location
                            Application.References.Remove ref
                            Set ref = Nothing
                            Application.References.AddFromFile strFile
                            Debug.Print "Updated Version Control System reference to: " & ref.FullPath
                            blnInitialize = True
                            blnLoaded = True
                        Else
                            strError = "Could not find library file in " & strFile
                        End If
                        Exit For
                    Else
                        ' Found existing reference
                        blnInitialize = True
                        blnLoaded = True
                    End If
                Else
                    ' Disable version control
                    Run strLibPrefix & "ReleaseObjectReferences"
                    Application.References.Remove ref
                    Set ref = Nothing
                    Debug.Print "Removed Version Control System"
                    Exit For
                End If
            End If
        End If
    Next intCnt
    
    
    ' See if this module is being used as the VCS loader.
    If CodeProject.Name <> strLibraryFile Then
        
        ' Make sure library is loaded.
        If blnUseVersionControl And Not blnLoaded Then
            
            ' Attempt to load the file
            If strLibraryPath <> "\" And Dir(strLibraryPath, vbDirectory) <> "" Then
                ' Use specified path
                strPath = strLibraryPath
            Else
                ' Use current folder
                strPath = CodeProject.Path & "\"
            End If
            
            ' Check to see if the library file exists
            strPath = strPath & strLibraryFile
            If Dir(strPath) <> "" Then
                ' File exists
                If strPath <> CodeProject.FullName Then
                    ' In ADP projects, the first call may not fully load the library.
                    ' (Not exactly sure why, but ADP seems to work a little differently with
                    '  library databases. If we try the AddFromFile again, it finishes the
                    '  load and allows us to continue with calling the VCS functions.)
                    On Error Resume Next
                    References.AddFromFile strPath
                    If Err Then
                        If Err.Number = 35021 Then
                            Err.Clear
                            ' Try loading it again
                            References.AddFromFile strPath
                            If Not Err Then
                                ' Go ahead and run a compile to just to make sure we can run the VCS command
                                ' later in this code. (Name conflicts would probably be the reasons for errors here.)
                                On Error GoTo 0
                                DoCmd.RunCommand acCmdCompileAllModules
                            End If
                        End If
                    End If
                    ' Check for errors and reset error handler
                    If Err Then
                        ' Unresolved error after attempting to reload.
                        strError = "Could not load " & strPath & vbCrLf & "Error Number: " & Err.Number & Err.Description
                        Err.Clear
                    Else
                        blnInitialize = True
                    End If
                    On Error GoTo 0
                End If
            End If
        End If
    Else
        ' Running from library (i.e. code development in the library project)
        blnInitialize = blnUseVersionControl
        If Not blnInitialize Then Run strLibPrefix & "ReleaseObjectReferences"
    End If
    
    If strError = "" Then
        ' Initialize the VBE menu
        ' (Use the Run commmand to avoid compile errors if the library was not loaded)
        If blnInitialize Then Run strLibPrefix & "LoadVersionControlMenu", colParams
    Else
        Debug.Print "**ERROR** " & strError
    End If
            
End Sub